src/Sarif/DelegatingStream.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    internal class DelegatingStream : Stream
+    {
+        private readonly Stream stream;
+
+        /// <summary>
+        /// This is a wrapper for a stream that protects the underlying stream from being disposed.
+        /// </summary>
+        /// <param name="underlyingStream"></param>
+        internal DelegatingStream(Stream underlyingStream)
+        {
+            stream = underlyingStream;
+            stream.Seek(0L, SeekOrigin.Begin);
```

Comment:
>Seek [](http://example.com/codeflow?start=19&length=4)

this will raise an exception if the stream isn't seekable.

Reply:
So. thinking this over, I suppose you can assume that a user will use the stream properly (i.e., understand if it isn't seekable) or otherwise get an exception when the wrapped methods are called.

You shouldn't seek here because you will break this class for all non-seekable streams. You need a unit test for this case, by the way.

Why, exactly, are you setting the stream pointer to 0? Why is that our job in this wrapper? Shouldn't it not try to interfere with the stream in any way? 

My suggestion is just delete this line, leave the rest alone, and add relevant tests.

Reply:
Delegated reading already does not support non seekable streams (https://github.com/microsoft/sarif-sdk/blob/ea44355584c161aec784ee9d2aa3c57f84875906/src/Sarif/Readers/DeferredDictionary.cs#L69) and an exception is thrown for the argument in this PR if you try to create a jsonpositionedtextreader with a non seekable stream.

Reply:
> So. thinking this over, I suppose you can assume that a user will use the stream properly (i.e., understand if it isn't seekable) or otherwise get an exception when the wrapped methods are called.
> 
> You shouldn't seek here because you will break this class for all non-seekable streams. You need a unit test for this case, by the way.

I understand the concern now. I made the class internal and it was specifically designed for use with JsonPositionedTextReader - which does not support non-seekable streams in the first place.  The constructor calls LineMappingStreamReader.ctor which calls FindBomWidth which seeks.  You can verify this by getting rid of the seek in the ctor and throwing an exception on the overloaded seek method.

> 
> Why, exactly, are you setting the stream pointer to 0? Why is that our job in this wrapper? Shouldn't it not try to interfere with the stream in any way?

The wrapper is a contribution from another project where we used it as the workaround for the current behavior. In that project we wanted to read from 0 and we always knew the source of our stream was seekable. I think you're right that here it isn't needed.

> 
> My suggestion is just delete this line, leave the rest alone, and add relevant tests.